python/fastfec.py

```diff
@@ -1,59 +1,386 @@
+
+        * parse: parse the input file line-by-line
+
+            with open('file.fec', 'rb') as f:
+                for form, line in fastfec.parse(f):
+                    print("Got line:", form, line)
+
+        * parse_as_files: parse the input file into output files
+
+            import os
+            from pathlib import Path
+
+            # Custom open method
+            def open_output_file(filename, *args, **kwargs):
+                filename = os.path.join('custom_output_dir', filename)
+                output_file = Path(filename)
+                output_file.parent.mkdir(exist_ok=True, parents=True)
+                return open(filename, *args, **kwargs)
+
+            with open('file.fec', 'rb') as f:
+                fastfec.parse_as_files(f, open=open_output_file)
+
+        Close the library:
+
+            fastfec.free()
+        """
+
+    def __init__(self):
+        self.__init_lib()
 
         # Initialize
-        self.file_descriptors = {}
-        self.last_filename = None
-        self.last_fd = None
-        self.bytes_written = 0
-        self.buffer_read_fn = BUFFER_READ(
-            make_read_buffer(file_handle))
-        write_callback = self.provide_write_callback(output_directory)
-        self.write_callback_fn = CUSTOM_WRITE(write_callback)
-
-        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext(
-            1)
-        self.fec_context = self.libfastfec.newFecContext(
-            self.persistent_memory_context, self.buffer_read_fn, BUFFER_SIZE, self.write_callback_fn, BUFFER_SIZE, None, None, None, 0, 1)
-
-    def init_lib(self):
-        # TODO: ensure this works on any platform
-        self.libfastfec = CDLL(os.path.join(SCRIPT_DIR, "../zig-out/lib/libfastfec.dylib"))
-
-        # Lay out arg/res types
+        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext()
+
+    def parse(self, file_handle):
+        """Parses the input file line-by-line
+
+        Arguments:
+            file_handle -- An input stream for reading a .fec file
+
+        Returns:
+            An iterator that receives the form name and a dictionary
+            object describing each line in the file"""
+        # Set up a queue to be able to yield output
+        queue = Queue()
+        done_processing = object()
+
+        # Provide a custom line callback
+        buffer_read_fn = provide_read_callback(file_handle)
+        line_callback_fn = CUSTOM_LINE(
+            provide_line_callback(queue))
+        fec_context = self.libfastfec.newFecContext(
+            self.persistent_memory_context, buffer_read_fn, BUFFER_SIZE, CUSTOM_WRITE(
+                0),
+            BUFFER_SIZE, line_callback_fn, 0, None, None, None, 0, 1, 0)
+
+        # Run the parsing in a separate thread
+        def task():
+            self.libfastfec.parseFec(fec_context)
+            queue.put(done_processing)
+
+        Thread(target=task, args=()).start()
```

Comment:
Is this just single threaded in a different thread? Could we use multiple if they are available?

Reply:
It's basically single-threaded in a different way. I added more context in the comments

Reply:
Multi-threading wouldn't work unfortunately